js/ui/components/clubhouse.js

```diff
@@ -0,0 +1,323 @@
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+//
+// Author: Joaquim Rocha <jrocha@endlessm.com>
+//
+
+const Clutter = imports.gi.Clutter;
+const Flatpak = imports.gi.Flatpak
+const Gio = imports.gi.Gio;
+const GLib = imports.gi.GLib;
+const GObject = imports.gi.GObject;
+const Shell = imports.gi.Shell;
+const Signals = imports.signals;
+const St = imports.gi.St;
+
+const GnomeSession = imports.misc.gnomeSession;
+const Lang = imports.lang;
+const Main = imports.ui.main;
+const MessageList = imports.ui.messageList;
+const MessageTray = imports.ui.messageTray;
+const Monitor = imports.ui.monitor;
+const NotificationDaemon = imports.ui.notificationDaemon;
+
+const sessionBus = Gio.bus_get_sync(Gio.BusType.SESSION, null);
+
+const CLUBHOUSE_ID = 'com.endlessm.Clubhouse';
+const CLUBHOUSE_DBUS_OBJ_PATH = '/com/endlessm/Clubhouse';
+
+var ClubhouseNotificationBanner = new Lang.Class({
+    Name: 'ClubhouseNotificationBanner',
+    Extends: MessageTray.NotificationBanner,
+
+    _init: function(notification) {
+        this.parent(notification);
+
+        // We don't have an "unexpanded" state for now
+        this.expanded = true;
+        this._actionBin.visible = true;
+
+        this.actor.can_focus = true;
+        this.actor.track_hover = true;
+
+        // Override the style name because this is a not a regular notification
+        this.actor.remove_style_class_name('notification-banner');
+        this.actor.add_style_class_name('clubhouse-quest-view');
+        this._iconBin.add_style_class_name('clubhouse-quest-view-icon-bin');
+
+        // Always wrap the body's text
+        this.bodyLabel.actor.clutter_text.line_wrap = true;
+        this.bodyLabel.actor.add_style_class_name('clubhouse-quest-view-label');
+
+        this._closeButton.connect('clicked', Lang.bind(this, function() {
+            this.emit('dismiss');
+        }));
+    },
+});
+Signals.addSignalMethods(ClubhouseNotificationBanner.prototype);
+
+var ClubhouseNotification = new Lang.Class({
+    Name: 'ClubhouseNotification',
+    Extends: NotificationDaemon.GtkNotificationDaemonNotification,
+
+    _init: function(source, notification) {
+        this._currentIconPath = null;
+
+        let gicon = notification.icon;
+        if (gicon) {
+            let newGicon = this._getClubhouseIcon(Gio.icon_deserialize(gicon));
```

Comment:
> I'm not entirely sure what this code is trying to do, but there are a few problems with this approach:

I am using a `GFileIcon` from the Clubhouse side. This [serializes the content as the URI](https://gitlab.gnome.org/GNOME/glib/blob/master/gio/gfileicon.c#L259).
I implemented it this way because if we use use the real icon's data we will be sending that over D-Bus all the time (even if we cache stuff in the Shell side). So I thought it'd be more optimal to just send the icon URI, but since the icons are shipped with the Clubhouse Flatpak, I have to look into it.
I know probably the optimization is negligible here, so if you really prefer to send data all the time I can do it.

> * it feels a bit awkward in general to have the shell look into the clubhouse Flatpak to find these icons

I understand but this is how I managed to skip sending the icons' data all the time.


> * if we need to have the shell look at the clubhouse Flatpak, the current code looks into the Flatpak user installation to locate icons, but the clubhouse will be installed system-wide

That's a mistake, I intended to add the counterpart check but forget to do it. Thanks for spotting it.

> * `this._getClubhouseIcon()` can return `null` but the code here does not account for that

I will take care of that.

Reply:
I will keep the current approach of using `FileIcons` and looking into the app for now. If you still disagree after my explanation, we can change it.

Reply:
@joaquimrocha I understand why it was done like this, but I do feel that it would be cleaner to send the image data over the wire for now; as long as it's done once per-notification the overhead should not be too bad as these are not large images. We can of course revisit it if it turns out to be a problem. I'm also keen on avoiding a dependency between the Shell and Flatpak for now.

Reply:
Okay. I have changed that.